### PR TITLE
Specify States of RNN Layers Symbolically

### PR DESCRIPTION
**Pull request by Joshua-Chin**
This adds `specify_states` flag to the `Recurrent` Layer and modifies the `BasicRNN`, `GRU`, and `LSTM` layers to use the flag. This flag allows you to call the recurrent layer on a list of tensors. The first element of the list is the input tensor. The remaining elements are the initial state tensors.

See #5358.

### CODE REVIEW

**Comment by fchollet** [comment]
This is too complicated. We don't need a new argument in the constructor API. We just need an `initial_state` argument in `call`. Can be a tensor or list of tensors (like masks, basically). It's just a few lines of code, the main thing is to write something about it in the docstring (with examples), and the demonstrate it in a seq2seq example script in the examples folder.

**Comment by Joshua-Chin** [comment]
If we input `initial_state` as a keyword argument of `call`, the layer that outputted `initial_state` will not be recorded in `inbound_layers`.  In the case of a seq2seq model, where the encoder and decoder have distinct weights, only the decoder would be trained. If we treat `initial_state` as an input layer, the weights of the encoder will be properly recorded.

**Comment by fchollet** [comment]
Good point, that's right. It's unfortunate that a system that's purely in the background (nodes) would force us to make API choices, though.

We can either think of ways to retain layer connectivity with a tensor keyword argument in `call`, or we can set `call` to take either a single tensor or a list of 2 tensors `(inputs, initial_states)`.

**Comment by Joshua-Chin** [comment]
This latest commit modifies `call` to take either a single tensor or a list of tensors. When given a list of tensors, the first element is `inputs`, and the remaining tensors are `initial_states` (`LSTM` has a tuple of tensors as its state). This commit also changes the `input_spec` of `Recurrent` to `None` so that `call` can accept a variable number of tensors.

**Comment by farizrahman4u** [comment]
Accepting a list of tensors sounds good.Another option would be to accept `dict` inputs, like `y = LSTM({'input': x, 'states': [h, c]})`

**Comment by fchollet** [comment]
Accepting a list of 2 tensors is basically our only option when using as state the output of a Keras layer. However it doesn't cover the case where the states don't come from a Keras layer. You could always wrap them in an `InputLayer`, but it feels inelegant.

One solution (not saying we should do this, just putting it out there) is to accept a `initial_state` keyword argument, and to add a `__call__` method in RNN layers that would:

- Check if `initial_state` is a Keras tensor (has a Keras history)
- if so, put the state(s) in the list of inputs, and call the parent's `__call__`
- else, keep it as a keyword argument and call the parent's `__call__`

The `call` method would then be written so as to accept either an inputs list or a keyword argument `initial_state`.

That way the API would always be `RNN(inputs, initial_state=[x, y, z])` or `RNN(inputs, initial_state=x)` no matter where the states come from. This, at the cost of a somewhat more complicated RNN base class.

**Comment by fchollet** [comment]
After reflexion, I think we should go with above, simply because good API design takes priority over everything else (including implementation simplicity and elegance, which is important and great to have, but still secondary to user-facing API design).

Any thoughts?

**Comment by Joshua-Chin** [comment]
I agree that implementation simplicity is secondary to user-facing design. However, I have a few questions about the proposed API. Would it accept or reject `RNN(inputs, initial_state)` and `RNN([inputs, initial_state])`?

**Comment by fchollet** [comment]
> Would it accept or reject RNN(inputs, initial_state) and RNN([inputs, initial_state])

`RNN.call([inputs, state])` or `RNN.call([inputs, state1, state2])` would work. Essentially, `RNN.call(inputs, initial_state=[state1, state2])` would map to the former and would be the preferred, documented API.

**Comment by Joshua-Chin** [comment]
This commit implements the API `RNN.call(inputs, initial_state=[state1, state2])`. In addition, the `initial_state` can be validated by having `RNN.build` define an `RNN.state_spec`.

This commit currently rejects `RNN.call([inputs, state1, state2])`, because it makes the implementation simpler. However, it wouldn't be too difficult to modify the code to accept `RNN.call([inputs, state1, state2])`.

This code expects `initial_state` to be either a tensor or list of tensors. It appears to me that specify `initial_state` as a numpy array would not be very useful. Because we are specifying the `initial_state` in `call`, using a numpy array would cause the `initial_state` to be constant over all inputs. In addition, the code does not explicitly check if the tensors are  keras tensors. If it did, it might cause problems when using keras as a layer library for Tensorflow / Theano.

**Comment by fchollet** [comment]
Good job, thanks. I merged it with some modifications.  https://github.com/fchollet/keras/commit/da21c15180d12a74a7c744c1f851a94c8b931a63

**Comment by fchollet** [comment]
If you're interested, the next step is to write a canonical example of how to set up seq2seq in Keras, and to add it as a new script in `examples/`. I won't have time to do it for a pretty long while, so feel free to take care of it!
